File: app/src/main/java/com/project/scratchstudio/kith_andoid/Service/PhotoService.java
package com.project.scratchstudio.kith_andoid.Service;

import android.app.Activity;
import android.content.res.Resources;
import android.database.Cursor;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Matrix;
import android.media.ExifInterface;
import android.net.Uri;
import android.provider.MediaStore;
import android.util.Base64;
import android.util.Log;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.UnsupportedEncodingException;
import java.net.HttpURLConnection;
import java.net.URL;

public class PhotoService {

    private Activity context;

    public PhotoService(Activity context) {
        this.context = context;
    }

    String base64Photo(Bitmap photo) throws UnsupportedEncodingException {
        if (photo == null) {
            return null;
        }

        ByteArrayOutputStream byteArrayOS = new ByteArrayOutputStream();
        photo.compress(Bitmap.CompressFormat.JPEG, 30, byteArrayOS);
        byte[] result = Base64.encode(byteArrayOS.toByteArray(), Base64.NO_WRAP);
        return new String(result, "UTF-8");
    }

    public Bitmap resizeBitmap(Bitmap bm, int newWidth, int newHeight) {
        int width = bm.getWidth();
        int height = bm.getHeight();
        float scaleWidth = ((float) newWidth) / width;
        float scaleHeight = ((float) newHeight) / height;
        Matrix matrix = new Matrix();
        matrix.postScale(scaleWidth, scaleHeight);
        Bitmap resizedBitmap = Bitmap.createBitmap(bm, 0, 0, width, height, matrix, false);
        bm.recycle();
        return resizedBitmap;
    }

    public Bitmap compressPhoto(Bitmap bitmap) {
        File file = new File(context.getFilesDir(), "tmp");
        if (!file.exists()) {
            file.mkdirs();
        }
        File img = new File(file, "tmp_img");
        try {
            FileOutputStream out = new FileOutputStream(img);
            bitmap.compress(Bitmap.CompressFormat.JPEG, 50, out);

        } catch (IOException e) {
            e.printStackTrace();
        }
        context.deleteFile("tmp_img");
        return bitmap;
    }


    private int calculateInSampleSize(BitmapFactory.Options options, int reqWidth, int reqHeight) {
        final int height = options.outHeight;
        final int width = options.outWidth;
        int inSampleSize = 1;

        if (height > reqHeight || width > reqWidth) {

            final int halfHeight = height / 2;
            final int halfWidth = width / 2;

            while ((halfHeight / inSampleSize) >= reqHeight && (halfWidth / inSampleSize) >= reqWidth) {
                inSampleSize *= 2;
            }
        }

        return inSampleSize;
    }

    public Bitmap changePhoto(Bitmap bitmap, Uri uri) {
        int angel = -1;
        String path = getRealPathFromURI(uri);
        File f = new File(path);
        try {
            angel = getExifAngle(f);
        } catch (Exception e) {
            Log.i("ERROR IN ANGEL:", e.getMessage());
        }
        bitmap = changeOrientation(f, bitmap, angel);
        return bitmap;
    }

    private Bitmap changeOrientation(File f, Bitmap bitmap, int angle) {
        Matrix mat = new Matrix();
        mat.postRotate(angle);
        BitmapFactory.Options options = new BitmapFactory.Options();
        options.inSampleSize = 2;

        try {
            Bitmap bmp = BitmapFactory.decodeStream(new FileInputStream(f),
                    null, options);
            bitmap = Bitmap.createBitmap(bmp, 0, 0, bmp.getWidth(),
                    bmp.getHeight(), mat, true);
            ByteArrayOutputStream outstudentstreamOutputStream = new ByteArrayOutputStream();
            bitmap.compress(Bitmap.CompressFormat.PNG, 100,
                    outstudentstreamOutputStream);
        } catch (IOException e) {
            Log.w("TAG", "-- Error in setting image");
        }

        return bitmap;
    }

    private int getExifAngle(File f) {
        int angle = 0;
        try {
            ExifInterface exif = new ExifInterface(f.getPath());
            int orientation = exif.getAttributeInt(
                    ExifInterface.TAG_ORIENTATION,
                    ExifInterface.ORIENTATION_NORMAL);

            if (orientation == ExifInterface.ORIENTATION_ROTATE_90) {
                angle = 90;
            } else if (orientation == ExifInterface.ORIENTATION_ROTATE_180) {
                angle = 180;
            } else if (orientation == ExifInterface.ORIENTATION_ROTATE_270) {
                angle = 270;
            }
        } catch (IOException e) {
            Log.w("TAG", "-- Error in setting image");
        } catch (OutOfMemoryError oom) {
            Log.w("TAG", "-- OOM Error in setting image");
        }

        return angle;
    }

    private String getRealPathFromURI(Uri contentURI) {

        String result;
        Cursor cursor = context.getContentResolver().query(contentURI, null, null, null, null);
        if (cursor == null) { // Source is Dropbox or other similar local file path
            result = contentURI.getPath();
        } else {
            cursor.moveToFirst();
            int idx = cursor.getColumnIndex(MediaStore.Images.ImageColumns.DATA);
            result = cursor.getString(idx);
            cursor.close();
        }
        return result;
    }

    public Bitmap decodeBase64Image(String encodedImage) throws UnsupportedEncodingException {
        byte[] decodedString = Base64.decode(encodedImage.getBytes("UTF-8"), Base64.DEFAULT);
        return BitmapFactory.decodeByteArray(decodedString, 0, decodedString.length);
    }

    public Bitmap getBitmapFromURL(String src) {
        try {
            URL url = new URL(src);
            HttpURLConnection connection = (HttpURLConnection) url.openConnection();
            connection.setDoInput(true);
            connection.connect();
            InputStream input = connection.getInputStream();
            Bitmap myBitmap = BitmapFactory.decodeStream(input);
            return myBitmap;
        } catch (IOException e) {
            // Log exception
            return null;
        }
    }

    public Uri getImageUri(Bitmap inImage) {
        ByteArrayOutputStream bytes = new ByteArrayOutputStream();
        inImage.compress(Bitmap.CompressFormat.JPEG, 100, bytes);
        String path = MediaStore.Images.Media.insertImage(context.getContentResolver(), inImage, "Title", null);
        return Uri.parse(path);
    }

    public Bitmap decodingPhoto(Resources res, int id, int width, int height) {
        BitmapFactory.Options options = new BitmapFactory.Options();
        options.inJustDecodeBounds = true;
        BitmapFactory.decodeResource(res, id, options);
        options.inSampleSize = calculateInSampleSize(options, width, height);
        options.inJustDecodeBounds = false;
        Bitmap resultBitmap = BitmapFactory.decodeResource(res, id, options);
        return resultBitmap;
    }
}

File: app/src/main/java/com/project/scratchstudio/kith_andoid/utils/PhotoService.kt
package com.project.scratchstudio.kith_andoid.utils

import android.content.Context
import android.content.res.Resources
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Matrix
import android.media.ExifInterface
import android.net.Uri
import android.provider.MediaStore
import android.util.Base64
import android.util.Log
import com.project.scratchstudio.kith_andoid.model.PhotoModelView

import java.io.ByteArrayOutputStream
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.io.UnsupportedEncodingException
import java.net.HttpURLConnection
import java.net.URL
import java.nio.charset.StandardCharsets.UTF_8
import java.util.Calendar

class PhotoService(private val context: Context) {

    fun fullPreparingPhoto(bitmap: Bitmap, uri: Uri): PhotoModelView? {
        val newPhoto = PhotoModelView()
        val newBitmap: Bitmap
        val path = getRealPathFromURI(uri)
        val f = File(path!!)

        newBitmap = changeOrientation(f, bitmap, getExifAngle(f))

        val file = File(context.filesDir, "kith")
        if (!file.exists()) {
            file.mkdirs()
        }
        val img = File(file, "kith_img${Calendar.getInstance().time}")
        val out = FileOutputStream(img)
        newBitmap.compress(Bitmap.CompressFormat.JPEG, 50, out)

        newPhoto.photoBitmap = newBitmap
        newPhoto.photoPhoneStoragePath = img.path
        newPhoto.phoneStorageFile = img

        return newPhoto
    }

    fun compressPhoto(bitmap: Bitmap): Bitmap {
        val file = File(context.filesDir, "tmp")
        if (!file.exists()) {
            file.mkdirs()
        }
        val img = File(file, "tmp_img")
        try {
            val out = FileOutputStream(img)
            bitmap.compress(Bitmap.CompressFormat.JPEG, 50, out)

        } catch (e: IOException) {
            e.printStackTrace()
        }

        context.deleteFile("tmp_img")
        return bitmap
    }

    fun changePhoto(bitmap: Bitmap, uri: Uri): Bitmap {
        var angel = -1
        val path = getRealPathFromURI(uri)
        val f = File(path)
        try {
            angel = getExifAngle(f)
        } catch (e: Exception) {
            Log.i("ERROR IN ANGEL:", e.message)
        }
        return changeOrientation(f, bitmap, angel)
    }

    fun changePhoto(bitmap: Bitmap, file: File): Bitmap {
        var angel = -1
        try {
            angel = getExifAngle(file)
        } catch (e: Exception) {
            Log.i("ERROR IN ANGEL:", e.message)
        }
        return changeOrientation(file, bitmap, angel)
    }

    private fun changeOrientation(f: File, bitmap: Bitmap, angle: Int): Bitmap {
        var newBitmap = bitmap
        val mat = Matrix()
        mat.postRotate(angle.toFloat())
        val options = BitmapFactory.Options()
        options.inSampleSize = 2

        try {
            val bmp = BitmapFactory.decodeStream(FileInputStream(f), null, options)
            newBitmap = Bitmap.createBitmap(bmp!!, 0, 0, bmp.width,
                    bmp.height, mat, true)
            val outstudentstreamOutputStream = ByteArrayOutputStream()
            newBitmap.compress(Bitmap.CompressFormat.PNG, 100,
                    outstudentstreamOutputStream)
        } catch (e: IOException) {
            Log.w("TAG", "-- Error in setting image")
        }

        return newBitmap
    }

    private fun getExifAngle(f: File): Int {
        var angle = 0
        try {
            val exif = ExifInterface(f.path)
            val orientation = exif.getAttributeInt(
                    ExifInterface.TAG_ORIENTATION,
                    ExifInterface.ORIENTATION_NORMAL)

            if (orientation == ExifInterface.ORIENTATION_ROTATE_90) {
                angle = 90
            } else if (orientation == ExifInterface.ORIENTATION_ROTATE_180) {
                angle = 180
            } else if (orientation == ExifInterface.ORIENTATION_ROTATE_270) {
                angle = 270
            }
        } catch (e: IOException) {
            Log.w("TAG", "-- Error in setting image")
        } catch (oom: OutOfMemoryError) {
            Log.w("TAG", "-- OOM Error in setting image")
        }

        return angle
    }

    fun getRealPathFromURI(contentURI: Uri): String? {
        val result: String?
        val cursor = context.contentResolver.query(contentURI, null, null, null, null)
        if (cursor == null) { // Source is Dropbox or other similar local file path
            result = contentURI.path
        } else {
            cursor.moveToFirst()
            val idx = cursor.getColumnIndex(MediaStore.Images.ImageColumns.DATA)
            result = cursor.getString(idx)
            cursor.close()
        }
        return result
    }

    private fun calculateInSampleSize(options: BitmapFactory.Options, reqWidth: Int, reqHeight: Int): Int {
        val height = options.outHeight
        val width = options.outWidth
        var inSampleSize = 1

        if (height > reqHeight || width > reqWidth) {

            val halfHeight = height / 2
            val halfWidth = width / 2

            while (halfHeight / inSampleSize >= reqHeight && halfWidth / inSampleSize >= reqWidth) {
                inSampleSize *= 2
            }
        }

        return inSampleSize
    }

    @Throws(UnsupportedEncodingException::class)
    fun base64Photo(photo: Bitmap?): String {
        if (photo == null) {
            return ""
        }

        val byteArrayOS = ByteArrayOutputStream()
        photo.compress(Bitmap.CompressFormat.JPEG, 30, byteArrayOS)
        val result = Base64.encode(byteArrayOS.toByteArray(), Base64.NO_WRAP)
        return String(result, UTF_8)
    }

    fun resizeBitmap(bm: Bitmap, newWidth: Int, newHeight: Int): Bitmap {
        val width = bm.width
        val height = bm.height
        val scaleWidth = newWidth.toFloat() / width
        val scaleHeight = newHeight.toFloat() / height
        val matrix = Matrix()
        matrix.postScale(scaleWidth, scaleHeight)
        val resizedBitmap = Bitmap.createBitmap(bm, 0, 0, width, height, matrix, false)
        bm.recycle()
        return resizedBitmap
    }

    @Throws(UnsupportedEncodingException::class)
    fun decodeBase64Image(encodedImage: String): Bitmap {
        val decodedString = Base64.decode(encodedImage.toByteArray(charset("UTF-8")), Base64.DEFAULT)
        return BitmapFactory.decodeByteArray(decodedString, 0, decodedString.size)
    }

    fun getBitmapFromURL(src: String): Bitmap? {
        try {
            val url = URL(src)
            val connection = url.openConnection() as HttpURLConnection
            connection.doInput = true
            connection.connect()
            val input = connection.inputStream
            return BitmapFactory.decodeStream(input)
        } catch (e: IOException) {
            // Log exception
            return null
        }

    }

    fun getImageUri(inImage: Bitmap): Uri {
        val bytes = ByteArrayOutputStream()
        inImage.compress(Bitmap.CompressFormat.JPEG, 100, bytes)
        val path = MediaStore.Images.Media.insertImage(context.contentResolver, inImage, "Title", null)
        return Uri.parse(path)
    }

    fun decodingPhoto(res: Resources, id: Int, width: Int, height: Int): Bitmap {
        val options = BitmapFactory.Options()
        options.inJustDecodeBounds = true
        BitmapFactory.decodeResource(res, id, options)
        options.inSampleSize = calculateInSampleSize(options, width, height)
        options.inJustDecodeBounds = false
        return BitmapFactory.decodeResource(res, id, options)
    }
}
